decisions/0015-separate-library-per-product.md

```diff
@@ -0,0 +1,72 @@
+  bloating the common package but will start "polluting" our frontends with request/response models
+  that we're trying to hide behind services. Cognitive load is also not particularly improved.
+  Services that are used in more than 1 client would automatically end up in `common`.
+- **Create new libs for each product** - This involves splitting our common code into `common`,
+  `password-manager`, `admin-console` and `secrets-manager` library packages. It has all the benfits
+  of the `common` package such as being able to keep internal request/response models while avoiding
+  bloat and lowering cognitive load. It also continues to encourage separation of UI and domain
+  knowledge. The code inside the packages would be organized by feature.
+- **Create new libs for each feature** - Shares many similarities with the option above, but instead
+  splits the common code directly into features. Just like above, this has all the benfits of the
+  `common` package such as being able to keep internal request/response models while avoiding bloat
+  and lowering cognitive load. It also continues to encourage separation of UI and domain knowledge.
+
+## Decision Outcome
+
+Chosen option: **Create new libs for each product**
+
+### Positive Consequences
+
+- Write business logic without involving platform specifics.
+- Packages work cross platform and are easy to import into another client.
+- Ability to restrict access to internal classes and implementations.
+- Smaller packages usually result in less cognitive load.
+- Matches our organizational structure which in turn can be interpreted as matching our business
+  capabilities
+  [Amazon Whitepaper: Organized Around Business Capabilities](https://docs.aws.amazon.com/whitepapers/latest/running-containerized-microservices/organized-around-business-capabilities.html).
+
+### Negative Consequences
+
+- Less obvious where classes should be added. The real world rarely wants to fit neatly into boxes
+  and making smaller boxes usually means it is going to be harder to fit things into them.
+- Products that appear organically will not match structure resulting in friction during
+  development.
+
+### Product-pacakges vs. Feature-packages
+
+Creating one library per feature will result in smaller boxes. The biggest advantage of this is
+probably much smaller boxes and more fine-grained control over what gets imported where. However
+this also leads to "what is a feature?" becoming one of the most prominent questions, something that
+might not be as important in the other alternatives. Features have a tendency to get represented as
+"entities" which makes everything more complicated because "entities" have different meanings in
+different contexts. An "organization" probably means different things if you're inside an admin
+console vs. simply browsing passwords. Having only one feature for both cases will end up creating
+objects that contain to much for certain contexts.
```

Comment:
On the other hand this becomes frustrating when different contexts have the *same* way of looking at certain things. Eg. a user is probably a user not matter the context, though admin-console might have services that allow for more operations on a user.

Reply:
Closed and moved to https://github.com/bitwarden/contributing-docs/pull/12/